### DISCUSSION TITLE
Ignore Build Step: test not just the latest commit, but all commits since last deploy

### DISCUSSION THREAD

**Discussion by msafi** [vercel/community#1304] [Help] [answered]
If you use "Ignore Build Step" with this basic command...

```bash
git diff HEAD^ HEAD --quiet ./
```

...it only checks if the last commit has any changes. So if you have a local branch with 2 commits, with the first one having relevant changes, while the second (and latest one) doesn't. Then you push all of these commits to GitHub in one push, then Ignore Build Step will cancel the build because the latest commit doesn't have any changes. But it shouldn't do that because the first commit has relevant changes that need building.

The only solution I can think of is to test all of the commits on the branch since the last deployment. Is something like that possible? What do folks currently do to workaround this issue? Personally, I created a dummy file that I will edit when necessary to trigger a deployment.

**Comment by kylemh** [comment]
You can just check against the branch! I do wish this were on the "Ignored Build Step" docs page though!

`git diff $(git branch --show-current)^ HEAD --quiet ./`

To your point, this won't capture things _since the previous deployment_. You'd need to make a custom script using the Vercel API, but I at least find this better than each commit. I'd rather build too much, than miss the build.

**Reply by msafi** [reply]
Wouldn't that make it run unnecessarily sometimes? Like if you have a branch with 20 commits, with only the first commit having relevant changes to the project, then the build will still run for all the subsequent 19 commits which didn't have any changes that required building, right?

Edit: just saw your update. Good point :)

**Comment by gabrielalmeida** [comment]
We're facing the same issue. I assume it's a common scenario based on how popular monorepos are and Vercel supporting them natively. To my surprise, previous older discussions received little to no attention:

https://github.com/vercel/vercel/discussions/6087
https://github.com/vercel/vercel/discussions/7251

Exposing something like `$VERCEL_LAST_BUILD_SHA` on https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables probably would solve this.

**Reply by amyegan** [reply]
I think [`VERCEL_GIT_PREVIOUS_SHA`](https://vercel.com/changelog/september-2022-papercuts) is what you need! It was added a few months ago as a [System Environment Variable](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables)

**Comment by amyegan** [comment, marked as answer]
You can find more information about [Ignored Build Step](https://vercel.com/docs/concepts/projects/overview#ignored-build-step) in the feature documentation and in the guide on on [how to use the "Ignored Build Step" field on Vercel](https://vercel.com/guides/how-do-i-use-the-ignored-build-step-field-on-vercel).

The field can be customized with a more complex script using the [`VERCEL_GIT_PREVIOUS_SHA` environment variable](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) to compare with the last successful deployment. Note that the variable is only exposed when an Ignored Build Step is provided.

**Reply by rdsedmundo** [reply]
I'm facing something that could be related here @amyegan, would you clarify if I should open a new discussion for it?

I have just a single branch and I made a commit that started deploying, then I made a second commit that was ignored by the build step (correctly), but because a second commit was made the first deployment was canceled and it shouldn't have been.

**Reply by amyegan** [reply]
Hi @rdsedmundo! This sounds like a related, but slightly different problem. If you wouldn't mind opening a [new bug report](https://github.com/orgs/vercel/discussions/new?category=bug-report) post, I will share it with the rest of the team 🙏

**Reply by rdsedmundo** [reply]
Thanks for the quick answer, here it is: https://github.com/orgs/vercel/discussions/2431

**Comment by ibash** [comment]
`VERCEL_GIT_PREVIOUS_SHA` doesn't work as expected. In particular the git commit history isn't present, so you can't diff against it.

e.g. 

With an ignored build step of `git diff --quiet $VERCEL_GIT_PREVIOUS_SHA HEAD -- ./`

You get:
```
Cloning completed: 1.464s
--
16:52:18.089 | Running "git diff --quiet $VERCEL_GIT_PREVIOUS_SHA HEAD -- ./"
16:52:18.102 | fatal: bad object 783b175ffc68abffc1256ea95ce25d877f26226e
```

**Reply by rosslh** [reply]
You ever find a workaround for this?

**Reply by ibash** [reply]
Nope, I think I just stopped using ignore build step :/
